modules/cnn_3dobj/src/cnn_feature.cpp

```diff
@@ -0,0 +1,244 @@
+#include "precomp.hpp"
+using namespace caffe;
+using std::string;
+
+namespace cv
+{
+namespace cnn_3dobj
+{
+    descriptorExtractor::descriptorExtractor(const string& device_type, int device_id)
+    {
+        if (strcmp(device_type.c_str(), "CPU") == 0 || strcmp(device_type.c_str(), "GPU") == 0)
+        {
+            if (strcmp(device_type.c_str(), "CPU") == 0)
+            {
+                caffe::Caffe::set_mode(caffe::Caffe::CPU);
+                device_info.push_back("CPU");
+                std::cout << "Using CPU" << std::endl;
+            }
+            else
+            {
+                caffe::Caffe::set_mode(caffe::Caffe::GPU);
+                caffe::Caffe::SetDevice(device_id);
+                device_info.push_back("GPU");
+                std::cout << "Using GPU" << std::endl;
+                std::cout << "Using Device_id=" << device_id << std::endl;
+            }
+            net_set = true;
+        }
+        else
+        {
+            std::cout << "Error: Device name must be 'GPU' together with an device number or 'CPU'." << std::endl;
+            net_set = false;
+        }
+    };
+
+    std::vector<string> descriptorExtractor::getDevice()
+    {
+        std::vector<string> device_info_out;
+        device_info_out = device_info;
+        return device_info_out;
+    };
+
+    void descriptorExtractor::setDevice(const string& device_type, const string& device_id)
+    {
+        if (strcmp(device_type.c_str(), "CPU") == 0 || strcmp(device_type.c_str(), "GPU") == 0)
+        {
+            if (strcmp(device_type.c_str(), "CPU") == 0)
+            {
+                caffe::Caffe::set_mode(caffe::Caffe::CPU);
+                device_info.push_back("CPU");
```

Comment:
Multiple invocations to setDevice will increase the size of device_info? And also dev_id should not be saved in private variables?
Anyway because you are going to split the set and get methods, you can remove this vector and simply have a private variable for device type and another one for device_id